### PR TITLE
[Freeze][Do Not Modify] Fix #685 - Integrate DuckDB - E2E

### LINKED ISSUE #685: [Lake][ETL] DuckDB E2E - Ingestion -> Dashboards
### Motivation
We completed the techspikes around our data infrastructure. As an outcome, we're going to implement our ETL pipeline in DuckDB such that we maintain many of our constraints:
- We can test everything end-to-end (pytest, duckdb, plots)
- We can reduce requirements on servers/infra/ops by using in-memory, embedded, on-disk db (duckb)
- We can continue doing distributed computing (ray)
- DB store can grow elastic w/ lvm, efs, filestore

### Outline
**Our first goal is to take the current ETL workflow and update it end-to-end.**

Screenshot from 2024-02-29 13-51-46

### Shelved Deliverables
CLOSED TICKET - Add ETL checkpoint to enforce SLAs, and process data incrementally. [#694](https://github.com/oceanprotocol/pdr-backend/issues/694)

Reason: We're going to instead implement a build step that leverages a simple SQL strategy w/ temp tables, such that we can enforce SLAs in a clean manner.

## DoD
**[First Deliverable - Update Ingestion + Load]**
- [x] Update GQL local/save-to-disk to use csv [#681](https://github.com/oceanprotocol/pdr-backend/issues/681)
- [x] Fetch + insert new records into db [#682](https://github.com/oceanprotocol/pdr-backend/issues/682)

**[Core System Updates]**
- [x] Improve ETL by using SQL build strategy & temp tables to enforce SLAs [#771](https://github.com/oceanprotocol/pdr-backend/issues/771)

**[Update ETL Deliverables]**
- [x] DuckDB - Update existing logic that queries data from parquet using polars, to query_data from PersistentDataStore using SQL.
- [x] DuckDB - Provide unifying `etl_` view for ETL build steps. Downstream bronze and silver tables require data from both `live_` and `build_` tables. [#810](https://github.com/oceanprotocol/pdr-backend/issues/810)
- [x] DuckDB - Update ETL "bronze step" to use duckdb/sql [#738](https://github.com/oceanprotocol/pdr-backend/issues/738)
- [x] DuckDB - Clean up any final queries that fetch the whole database system. No select * from table_name anywhere. [#809](https://github.com/oceanprotocol/pdr-backend/issues/809)
- [x] DuckDB - Create logic to support an etl_view that joins both live + temp tables, making it easier to query data as it's being built [#810](https://github.com/oceanprotocol/pdr-backend/issues/810)
- [x] DuckDB - Port latest ETL "bronze slots" to use duckdb/sql [#740](https://github.com/oceanprotocol/pdr-backend/issues/740)
- [x] DuckDB - Extend test coverage to simulate multiple runs, and cover more edge cases. [#811](https://github.com/oceanprotocol/pdr-backend/issues/811)
- [x] DuckDB - Abstract [Prod->Temp->View] pattern [#881](https://github.com/oceanprotocol/pdr-backend/issues/881)
- [x] Accuracy - Update accuracy endpoint to use the ETL to get and store the data. [#854](https://github.com/oceanprotocol/pdr-backend/issues/854)
- [x] Streamlit - Integrate predictoor_income work and dashboards [#612](https://github.com/oceanprotocol/pdr-backend/issues/612)

**[ETL CLI Deliverables]**
- [x] Update ETL CLI and provide a clean interface that helps enforce SLAs. [#703](https://github.com/oceanprotocol/pdr-backend/issues/703)
- [x] Create `pdr analytics describe, query, validate, resume` CLI command [#883](https://github.com/oceanprotocol/pdr-backend/issues/883)

**[Cleanup Deliverables]**
- [x] Cleanup DuckDB implementation, remove hanging dependencies, remove table.df caching, switching from parquet to CSV, cleanup tests. [#737](https://github.com/oceanprotocol/pdr-backend/issues/737)
- [x] Clean truevalue and predvalue [#664](https://github.com/oceanprotocol/pdr-backend/pull/664)
- [x] Improve CSVDataStore and PersistentDataStore instantiation inside Table [#773](https://github.com/oceanprotocol/pdr-backend/issues/773)
- [x] Cleanup Tables, query them directly through CSVDataStore + PersistentDataStore [#772](https://github.com/oceanprotocol/pdr-backend/issues/772)
- [x] Remove reference to `parquet_dir` use `lake_dir` instead [#770](https://github.com/oceanprotocol/pdr-backend/issues/770)
- [x] PDS thread tests are failing, although CI/CD passes. [#941](https://github.com/oceanprotocol/pdr-backend/issues/941)
- [x] OHLCV + CSVDS has been reconciled [#858](https://github.com/oceanprotocol/pdr-backend/pull/858)
- [x] Lake cli command to update is implemented [#953](https://github.com/oceanprotocol/pdr-backend/issues/953)
- [x] Fix issue where CSV data exists, but raw table won't be recreated [#1038](https://github.com/oceanprotocol/pdr-backend/issues/1038)
- [x] Fix pdr-slots and quey logic that's causing the pipeline to break [#1036](https://github.com/oceanprotocol/pdr-backend/issues/1036)
- [x] Time functions are returning local timestamp and breaking the pipeline [#1070](https://github.com/oceanprotocol/pdr-backend/issues/1070)
- [x] Use max(values) to get the max timestamp across all duckdb tables [#1049](https://github.com/oceanprotocol/pdr-backend/issues/1049)
- [x] Expand validation tool to check/report duplicate rows [#1058](https://github.com/oceanprotocol/pdr-backend/issues/1058)
- [x] Update GQLDF to use the same end_ts for all subgraph queries. [#1068](https://github.com/oceanprotocol/pdr-backend/issues/1068) 
- [x] Stabilize Lake Validation & CLI Commands [#1036](https://github.com/oceanprotocol/pdr-backend/issues/1036)
- [x] Fix drop logic to use the proper timetamp, and not drop the whole table [#1079](https://github.com/oceanprotocol/pdr-backend/issues/1079)
- [x] Verify lake behavior is working as expected [#1000](https://github.com/oceanprotocol/pdr-backend/issues/1000)
- [x] DuckDB - Use bronze-predictions for checkpoint [#982](https://github.com/oceanprotocol/pdr-backend/issues/982) 
- [x] Expand ETL coverage to have multiple days of data (1mb files) and multiple operations. Enforce a benchmark performance against this.

**[Ratchet Integration]**
- [x] https://github.com/oceanprotocol/pdr-backend/issues/1109

**[Post-DuckDB Merge - Core Functionality]**
- [ ] Verify incremental table updates are working as expected [#1001](https://github.com/oceanprotocol/pdr-backend/issues/1001)
- [ ] Readme for working the lake and completing UX e2e is documented and works well end-to-end [#1002](https://github.com/oceanprotocol/pdr-backend/issues/1002)
- [ ] DuckDB - Implement "silver predictions" [#665](https://github.com/oceanprotocol/pdr-backend/pull/665)
- [ ] DuckDB - Port latest ETL "silver predictions" to use duckdb/sql + close old PR [#741](https://github.com/oceanprotocol/pdr-backend/issues/741)
- [ ] DuckDB - Silver predictions SQL PR [#848](https://github.com/oceanprotocol/pdr-backend/pull/848)

**[Post-DuckDB - Peripheral Functionality]**
These are frozen. Do not start/complete until DuckDB review/work is complete.
- [x] Update OHCLV Data Factory to use DataStores [#769](https://github.com/oceanprotocol/pdr-backend/issues/769)
- [x] TableRegistry looks redundant now. Deprecate it. [#1088](https://github.com/oceanprotocol/pdr-backend/issues/1088)
- [ ] PredictoorETL is handling st_ts and end_ts correctly [#1086](https://github.com/oceanprotocol/pdr-backend/issues/1086)
- [ ] Re-enable slots and subscription raw tables.
- [ ] DuckDB - Re-enable subscription table [#1085](https://github.com/oceanprotocol/pdr-backend/issues/1085)
- [ ] DuckDB - Re-enable bronze_slots tables - [#595](https://github.com/oceanprotocol/pdr-backend/issues/595)
- [ ] ETL - Cleanup payout, truevals, and revenue calculations - [#1183](https://github.com/oceanprotocol/pdr-backend/issues/1183)
- [ ] Data Store Objects - Please rename functions to use sql nomenclature: `fill` becomes `insert`, `override` becomes `upsert`
- [ ] OHLCV + CSVDS will be updated after DuckDB has been updated [#769](https://github.com/oceanprotocol/pdr-backend/issues/769)
- [ ] If you delete CSVs and rebuild RAW tables, you get duplicate data [#1087](https://github.com/oceanprotocol/pdr-backend/issues/1087)

### PR DESCRIPTION
**Pull request by idiom-bytes**
Fixes #685

### CODE REVIEW

File: pdr_backend/analytics/test/test_get_predictions_info.py [hunk 1]
```diff
@@ -5,15 +5,14 @@
 from pdr_backend.lake.table import Table
 from pdr_backend.analytics.get_predictions_info import get_predictions_info_main
 from pdr_backend.ppss.ppss import mock_ppss
+from pdr_backend.lake.test.conftest import _clean_up_persistent_data_store
```
Comment (calina-c) [inline pdr_backend/analytics/test/test_get_predictions_info.py:8]: it it's importable, it shouldn't start with an underscore. That is the convention for strictly module functions.

File: pdr_backend/lake/base_data_store.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from typing import Dict
+
+from enforce_typing import enforce_types
+
+
+class BaseDataStore:
+    @enforce_types
+    def __new__(cls, base_path: str):
+        if not hasattr(cls, "_instances"):
```
Comment (calina-c) [inline pdr_backend/lake/base_data_store.py:9]: this looks like some kind of singleton, but I think it needs more context/comments. And spacing between blocks and before the return statement.

File: pdr_backend/lake/csv_data_store.py [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+    def _get_folder_path(self, dataset_identifier: str) -> str:
+        """
+        Returns the folder path for the given dataset_identifier.
+        If the folder does not exist, it will be created.
+        @args:
+            dataset_identifier: str - identifier of the dataset
+        """
+        folder_path = os.path.join(self.base_path, dataset_identifier)
+        os.makedirs(folder_path, exist_ok=True)
+        return folder_path
```
Comment (calina-c) [inline pdr_backend/lake/csv_data_store.py:22]: please add a line before return statements (multiple instances in this file)

File: pdr_backend/lake/csv_data_store.py [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+
+    @enforce_types
+    def _fill_with_zero(self, number: int, length: int = 10) -> str:
```
Comment (calina-c) [inline pdr_backend/lake/csv_data_store.py:25]: this is not **filling**, it is **padding**. Please adjust the name of this function.
You can also use something MUCH more simple, e.g. zfill, rjust or number formatting with f:

https://stackoverflow.com/questions/339007/how-do-i-pad-a-string-with-zeros

**Review by calina-c — CHANGES_REQUESTED** [review]
I reviewed the first part of the changes. I'll add the rest to a subsequent review.

File: pdr_backend/lake/table_registry.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    _instance: Optional["TableRegistry"] = None
+    _tables: Dict[str, Table] = {}
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super(TableRegistry, cls).__new__(cls)
+        return cls._instance
+
+    @enforce_types
+    def register_table(self, table_name: str, table_args: Tuple[str, SchemaDict, PPSS]):
```
Comment (calina-c) [inline pdr_backend/lake/table_registry.py:19]: please add return types in this file

File: pdr_backend/lake/table_registry.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    @enforce_types
+    def get_tables(self, table_names: Optional[List] = None):
+        if table_names is None:
+            table_names = list(self._tables.keys())
+
+        target_tables = [self._tables.get(table_name) for table_name in table_names]
+
+        # do it this way to avoid returning None
+        target_tables = [table for table in target_tables if table is not None]
```
Comment (calina-c) [inline pdr_backend/lake/table_registry.py:42]: or simply filter out None values https://www.geeksforgeeks.org/filter-in-python/

File: pdr_backend/lake/test/resources.py [hunk 7]
```diff
@@ -54,18 +63,44 @@ def _filter_gql_tables_config(record_config: Dict, record_filter: str) -> Dict:
...
+
+def _clean_up_persistent_data_store(tmpdir, table_name=None):
+    # Clean up PDS
+    persistent_data_store = PersistentDataStore(str(tmpdir))
+
+    # Select tables from duckdb
+    table_names = persistent_data_store.get_table_names()
+
+    # Drop the tables
+    if table_name in table_names:
```
Comment (calina-c) [inline pdr_backend/lake/test/resources.py:96]: easier:
table_names = [table_name] if table_name else persistent_data_store.get_table_names()
for table in table_names:
             persistent_data_store.duckdb_conn.execute(f"DROP TABLE {table}")

File: pdr_backend/lake/test/test_plutil.py [hunk 8]
```diff
@@ -304,3 +305,18 @@ def test_text_to_df():
...
+    table_name = get_table_name("test")
+    assert table_name == "test"
+
+    table_name = get_table_name("test", TableType.TEMP)
+    assert table_name == "_temp_test"
+
+    table_name = get_table_name("test", TableType.NORMAL)
+    assert table_name == "test"
+
+    table_name = get_table_name("test", TableType.ETL)
```
Comment (calina-c) [inline pdr_backend/lake/test/test_plutil.py:321]: suggest to add: get_table_name("test", "inexistentType")

File: pdr_backend/lake/test/test_table.py [hunk 9]
```diff
@@ -1,241 +1,149 @@
...
 
+def _table_exists(persistent_data_store, table_name):
+    table_names = persistent_data_store.get_table_names()
+    return [table_name in table_names, table_name]
 
-def get_table_df(network, st_ut, fin_ut, config):
-    print(network, st_ut, fin_ut, config)
-    return pl.DataFrame([mocked_object], table_df_schema)
 
+def _clean_up_persistent_data_store(tmpdir, table_name):
```
Comment (calina-c) [inline pdr_backend/lake/test/test_table.py:18]: this seems like a violation of DRY with the other function doing the exact same thing in lake/test_resources

File: setup.py [hunk 10]
```diff
@@ -9,9 +9,11 @@
 # Installed by pip install pdr-backend
 # or pip install -e .
 install_requirements = [
+    "black==24.2.0",
     "bumpversion",
     "ccxt>=4.1.59",
     "coverage",
+    "duckdb",
```
Comment (calina-c) [inline setup.py:16]: please constrain the duckdb version.

**Review by calina-c — CHANGES_REQUESTED** [review]
2/2 done reviewing.

Looks like an awesome job! Most comments are style and readability-related suggestions. Please also remove print statements.

**Comment by KatunaNorbert** [comment]
I'm running into the following error when running the do_etl() flow in the bronz table step
Screenshot 2024-04-12 at 10 27 24

**Comment by KatunaNorbert** [comment]
fixed the type issue from above by specifying the data frame schema

**Comment by idiom-bytes** [comment]
As requested by @kdetry, please do not modify this PR any longer.

It's stable and we're working towards integrating it with main.

**Comment by idiom-bytes** [comment]
We have completed this merging this, closing the PR.
